chore(cli): Remove verbose osm mesh list output

johnsonshi opened the pull request:
This removes verbose output from the osm mesh list
command. The removed output includes osm mesh controller
pod output and supported smi versions.

In lieu of printing verbose output, the command
instead prints out instructions for the user
to obtain supported smi versions through cli commands.

Resolves #3648.

Signed-off-by: Johnson Shi <Johnson.Shi@microsoft.com>

<!--

Please mark with X for applicable areas.

-->
**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| Documentation              | [ ] |
| Install                    | [ ] |
| CLI Tool                   | [X] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [ ] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [ ] |


Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? N
    -   Did you notify the maintainers and provide attribution? NA

1. Is this a breaking change? N

johnsonshi commented:
Output:

```
~/dev/osm(remove-verbose-osm-mesh-list*) » ./bin/osm mesh list                                                                                                                                                      johnsonshi@Johnsons-MBP

MESH NAME   MESH NAMESPACE   VERSION   ADDED NAMESPACES
osm         osm-system       v0.9.0    test

To list the OSM controller pods, please run the following command
	kubectl get pods -n osm-system -l app=osm-controller

To list the supported SMI versions, please run the following command and obtain the supported versions from http://localhost:9091/smi/version
	kubectl port-forward -n osm-system $(kubectl get pods -n osm-system --selector app=osm-controller --no-headers | awk '{print $1}') 9091 &
	curl http://localhost:9091/smi/version
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(remove-verbose-osm-mesh-list*) » kubectl get pods -n osm-system -l app=osm-controller                                                                                                                     johnsonshi@Johnsons-MBP
NAME                              READY   STATUS    RESTARTS   AGE
osm-controller-7cc6bf7b44-mtz5z   1/1     Running   0          4d18h
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(remove-verbose-osm-mesh-list*) » kubectl port-forward -n osm-system $(kubectl get pods -n osm-system --selector app=osm-controller --no-headers | awk '{print $1}') 9091 &                                johnsonshi@Johnsons-MBP
[1] 3605
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(remove-verbose-osm-mesh-list*) » Forwarding from 127.0.0.1:9091 -> 9091                                                                                                                                   johnsonshi@Johnsons-MBP
Forwarding from [::1]:9091 -> 9091

--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/dev/osm(remove-verbose-osm-mesh-list*) » curl http://localhost:9091/smi/version                                                                                                                                   johnsonshi@Johnsons-MBP
Handling connection for 9091
{"HTTPRouteGroup":"specs.smi-spec.io/v1alpha4","TCPRoute":"specs.smi-spec.io/v1alpha4","TrafficSplit":"split.smi-spec.io/v1alpha2","TrafficTarget":"access.smi-spec.io/v1alpha3"}
```

shashankram commented:
> Output:
> 
> ```
> ~/dev/osm(remove-verbose-osm-mesh-list*) » ./bin/osm mesh list                                                                                                                                                      johnsonshi@Johnsons-MBP
> 
> MESH NAME   MESH NAMESPACE   VERSION   ADDED NAMESPACES
> osm         osm-system       v0.9.0    test
> 
> To list the OSM controller pods, please run the following command
> 	kubectl get pods -n osm-system -l app=osm-controller
> 
> To list the supported SMI versions, please run the following command and obtain the supported versions from http://localhost:9091/smi/version
> 	kubectl port-forward -n osm-system $(kubectl get pods -n osm-system --selector app=osm-controller --no-headers | awk '{print $1}') 9091 &
> 	curl http://localhost:9091/smi/version
> --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> ~/dev/osm(remove-verbose-osm-mesh-list*) » kubectl get pods -n osm-system -l app=osm-controller                                                                                                                     johnsonshi@Johnsons-MBP
> NAME                              READY   STATUS    RESTARTS   AGE
> osm-controller-7cc6bf7b44-mtz5z   1/1     Running   0          4d18h
> --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> ~/dev/osm(remove-verbose-osm-mesh-list*) » kubectl port-forward -n osm-system $(kubectl get pods -n osm-system --selector app=osm-controller --no-headers | awk '{print $1}') 9091 &                                johnsonshi@Johnsons-MBP
> [1] 3605
> --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> ~/dev/osm(remove-verbose-osm-mesh-list*) » Forwarding from 127.0.0.1:9091 -> 9091                                                                                                                                   johnsonshi@Johnsons-MBP
> Forwarding from [::1]:9091 -> 9091
> 
> --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> ~/dev/osm(remove-verbose-osm-mesh-list*) » curl http://localhost:9091/smi/version                                                                                                                                   johnsonshi@Johnsons-MBP
> Handling connection for 9091
> {"HTTPRouteGroup":"specs.smi-spec.io/v1alpha4","TCPRoute":"specs.smi-spec.io/v1alpha4","TrafficSplit":"split.smi-spec.io/v1alpha2","TrafficTarget":"access.smi-spec.io/v1alpha3"}
> ```

@johnsonshi, as an enhancement to this change, we could list the supported SMI policies below the table output, since the command to fetch these seems verbose in itself.

johnsonshi commented:
@shashankram Great idea!

jaellio commented on cmd/cli/mesh_list.go:
```diff
@@ -77,5 +71,13 @@ func (l *meshListCmd) run() error {
 	_ = w.Flush()
 
+	fmt.Fprintf(l.out, "\nTo list the OSM controller pods, please run the following command\n")
+	fmt.Fprintf(l.out, "\tkubectl get pods -n %s -l app=osm-controller\n", settings.Namespace())
```
Should the namespace for this command be generic ( <mesh namespace> ) since `settings.Namespace()` might not be the namespace of the mesh the user is trying to get the controller pods from? I think what you have above makes sense too

johnsonshi replied:
@jaellio Resolved - Instructed the user to instead put the namespace of the mesh they want to list.

```
	fmt.Fprintf(l.out, "\nTo list the OSM controller pods for a mesh, please run the following command passing in the mesh's namespace\n")
	fmt.Fprintf(l.out, "\tkubectl get pods -n <osm-mesh-namespace> -l app=osm-controller\n")
```

shashankram commented on cmd/cli/mesh_list.go:
```diff
@@ -77,5 +71,13 @@ func (l *meshListCmd) run() error {
+
+	supportedSmiVersionsEndpoint := fmt.Sprintf("http://localhost:%d%s", constants.OSMHTTPServerPort, constants.HTTPServerSmiVersionPath)
+	fmt.Fprintf(l.out, "\nTo list the supported SMI versions, please run the following command and obtain the supported versions from %s\n", supportedSmiVersionsEndpoint)
+	fmt.Fprintf(l.out, "\tkubectl port-forward -n %s $(kubectl get pods -n %s --selector app=osm-controller --no-headers | awk '{print $1}') %d &\n", settings.Namespace(), settings.Namespace(), constants.OSMHTTPServerPort)
+	fmt.Fprintf(l.out, "\tcurl %s\n", supportedSmiVersionsEndpoint)
```
2 issues here:
1. Address https://github.com/openservicemesh/osm/pull/3677#issuecomment-869828680
2. Not use linux specific commands such as `awk` in the output. The CLI is built for windows as well.

johnsonshi replied:
Resolved @shashankram. SMI info is now printed in a separate table.

johnsonshi commented:
This PR needs additional works and needs to be closed.

shashankram commented on cmd/cli/util.go:
```diff
@@ -54,19 +55,16 @@ func confirm(stdin io.Reader, stdout io.Writer, s string, tries int) (bool, erro
 	for _, meshInfo := range meshInfoList {
 		m := fmt.Sprintf(
-			"%s\t%s\t%s\t%s\t%s\t%s\n",
+			"%s\t%s\t%s\t%s\n",
```
In the future, we should use a pkg such as github.com/olekukonko/tablewriter for better formatting of tables, so long lines are wrapped around within the column.

shashankram reviewed (commented):
@johnsonshi could you share what the CLI output looks like with the latest change.

johnsonshi commented:
Sure thing @shashankram - The output looks like this

```
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » ./bin/osm mesh list                                                                                                                                                    johnsonshi@DESKTOP-AQ60TOB
No osm mesh control planes found
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » osm install                                                                                                                                                            johnsonshi@DESKTOP-AQ60TOB
OSM installed successfully in namespace [osm-system] with mesh name [osm]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » osm install --mesh-name osm-2 --osm-namespace osm-system-2                                                                                                             johnsonshi@DESKTOP-AQ60TOB
OSM installed successfully in namespace [osm-system-2] with mesh name [osm-2]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » kubectl create namespace test-1                                                                                                                                    1 ↵ johnsonshi@DESKTOP-AQ60TOB
namespace/test-1 created
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » kubectl create namespace test-2                                                                                                                                        johnsonshi@DESKTOP-AQ60TOB
namespace/test-2 created
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » osm namespace add test-1 --mesh-name osm --osm-namespace osm-system                                                                                                1 ↵ johnsonshi@DESKTOP-AQ60TOB
Namespace [test-1] successfully added to mesh [osm]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » osm namespace add test-2 --mesh-name osm-2 --osm-namespace osm-system-2                                                                                                johnsonshi@DESKTOP-AQ60TOB
Namespace [test-2] successfully added to mesh [osm-2]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » kubectl create namespace test-3                                                                                                                                        johnsonshi@DESKTOP-AQ60TOB
namespace/test-3 created
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » osm namespace add test-3 --mesh-name osm-2 --osm-namespace osm-system-2                                                                                                johnsonshi@DESKTOP-AQ60TOB
Namespace [test-3] successfully added to mesh [osm-2]
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
~/Dev/osm(remove-verbose-osm-mesh-list) » ./bin/osm mesh list                                                                                                                                                    johnsonshi@DESKTOP-AQ60TOB

MESH NAME   MESH NAMESPACE   VERSION   ADDED NAMESPACES
osm-2       osm-system-2     v0.9.1    test-2,test-3
osm         osm-system       v0.9.1    test-1

MESH NAME   MESH NAMESPACE   SMI SUPPORTED
osm-2       osm-system-2     HTTPRouteGroup:specs.smi-spec.io/v1alpha4,TCPRoute:specs.smi-spec.io/v1alpha4,TrafficSplit:split.smi-spec.io/v1alpha2,TrafficTarget:access.smi-spec.io/v1alpha3
osm         osm-system       HTTPRouteGroup:specs.smi-spec.io/v1alpha4,TCPRoute:specs.smi-spec.io/v1alpha4,TrafficSplit:split.smi-spec.io/v1alpha2,TrafficTarget:access.smi-spec.io/v1alpha3

To list the OSM controller pods for a mesh, please run the following command passing in the mesh's namespace
	kubectl get pods -n <osm-mesh-namespace> -l app=osm-controller
```

shashankram commented:
> HTTPRouteGroup:specs.smi-spec.io/v1alpha4,TCPRoute:specs.smi-spec.io/v1alpha4,TrafficSplit:split.smi-spec.io/v1alpha2,TrafficTarget:access.smi-spec.io/v1alpha3

@johnsonshi Would it make it more concise and less noisy by dropping the API group to make the above output look as follows:
`HTTPRouteGroup:v1alpha4, TCPRoute:v1alpha4, TrafficSplit:v1alpha2, TrafficTarget:v1alpha3`

johnsonshi commented on cmd/cli/util.go:
```diff
@@ -200,6 +234,16 @@ func getSupportedSmiForControllerPod(pod string, namespace string, restConfig *r
+		}
+	}
+
 	return smiSupported, nil
```
@shashankram With these changes, the API group is stripped out and only the version remains.

```
MESH NAME   MESH NAMESPACE   VERSION   ADDED NAMESPACES
osm         osm-system       v0.9.1    bookbuyer,bookstore,bookthief,bookwarehouse

MESH NAME   MESH NAMESPACE   SMI SUPPORTED
osm         osm-system       HTTPRouteGroup:v1alpha4,TCPRoute:v1alpha4,TrafficSplit:v1alpha2,TrafficTarget:v1alpha3

To list the OSM controller pods for a mesh, please run the following command passing in the mesh's namespace
	kubectl get pods -n <osm-mesh-namespace> -l app=osm-controller
```
